### PR TITLE
Refactoring Preprocessing

### PR DESCRIPTION
**Pull request by JeremieGauthier**
Laisse-moi savoir s'il y a des modifications à apporter. 

image

Si tu pouvais juste regarder pourquoi j'obtiens toujours `val_greedy_rouge_mean was not in top 3`, j'aimerais bien. Je ne voudrais pas que cette issue revienne durant le training sur CC.

### CODE REVIEW

File: src/domain/models/bertsum_transformer.py [hunk 1]
```diff
@@ -34,6 +47,10 @@ def forward(self, contents):
             contents["mark"],
             contents["mark_clss"],
         )
+        x = x.to(self.device)
+        segs = segs.to(self.device)
+        mask = mask.to(self.device)
```
Comment (TortillasAlfred) [inline src/domain/models/bertsum_transformer.py:52]: All those to() calls are not needed when i run the code. The DataLoader should automatically handle putting tensors on the right device. 

Also, if ever we want to go multi-GPU, this is the kind of behaviour that cannot be done when one has multiple GPUs.

File: src/domain/ucb.py [hunk 2]
```diff
@@ -87,3 +108,44 @@ def ucb(scorer, c_puct, n_samples, n_sents, priors=None):
...
+    for n in range(1, n_samples + 1):
+        ucb = q_vals + c_puct * priors * np.sqrt(2 * np.log(n) / n_visits)
+        ucb = np.nan_to_num(ucb, nan=np.inf)
+        threshold = np.partition(ucb, -3)[-3]
+        elligible_idxs = np.argwhere(ucb >= threshold)
+        elligible_idxs = elligible_idxs[:, 0]
+        sampled_idxs = np.random.choice(elligible_idxs, 3, replace=False)
+
+        sentence_gap = np.asarray(sentence_gap)
+        real_sampled_idxs = sampled_idxs + sentence_gap[sampled_idxs]
```
Comment (TortillasAlfred) [inline src/domain/ucb.py:130]: Very slick, well done.

**Review by TortillasAlfred — COMMENTED** [review]
Fantastic work. Correct the small details I pointed out and we will be able to merge this.

On a side note, I added a few minor commits of my own, notably to implement the idx_reparts portion in the forward() loop of the model. You already did all the heavy lifting, the code almost worked on its own.

File: src/domain/models/bertsum_transformer.py [hunk 3]
```diff
@@ -1,26 +1,39 @@
...
 
-        self.position_ids = torch.arange(MAX_LEN_DOCUMENT).to(self.device)
+        if MAX_LEN_DOCUMENT > 512:
+            pos_embeddings = nn.Embedding(MAX_LEN_DOCUMENT, self.bert.config.hidden_size)
+            bert_weight = self.bert.embeddings.position_embeddings.weight.data
+            bert_weight_dup = torch.cat((bert_weight,) * int(MAX_LEN_DOCUMENT // 512 + 1))
+            pos_embeddings.weight.data = bert_weight_dup[:MAX_LEN_DOCUMENT]
+            self.bert.embeddings.position_embeddings = pos_embeddings
 
+        self.encoder = TransformerInterEncoder(768, D_FFN, HEAD, DROPOUT, NUM_TLAYER)
```
[the remark this answers is not in the document]
Comment (JeremieGauthier) [reply src/domain/models/bertsum_transformer.py:35]: Nous ne devons pas changer ce paramètre. Il est à 768 par défaut dans le pretrained the bert-base-uncased

File: src/domain/models/bertsum_transformer.py [hunk 4]
```diff
@@ -1,26 +1,39 @@
...
+    def __init__(self, device, config):
         super(Summarizer, self).__init__()
         self.device = device
-        self.bert = BertModel.from_pretrained("bert-base-uncased").to(self.device)
-        self.encoder = TransformerInterEncoder(D_MODEL_BERT, D_FFN, HEAD, DROPOUT, NUM_TLAYER)
+        if config.bert_cache:
+            bert_cache_dir = join(getcwd(), "bert_cache/bertmodel_save_pretrained")
+            self.bert = BertModel.from_pretrained(bert_cache_dir, local_files_only=True).to(self.device)
+        else:
+            self.bert = BertModel.from_pretrained("bert-base-uncased").to(self.device)
```
[the remark this answers is not in the document]
Comment (JeremieGauthier) [reply src/domain/models/bertsum_transformer.py:26]: Oui, très bonne idée Mathieu!

[the remark this answers is not in the document]
Comment (JeremieGauthier) [reply src/domain/models/bertsum_transformer.py:26]: Je vais le changer immédiatement!

File: src/domain/models/bertsum_transformer.py [hunk 5]
```diff
@@ -1,26 +1,39 @@
...
 
-        self.position_ids = torch.arange(MAX_LEN_DOCUMENT).to(self.device)
+        if MAX_LEN_DOCUMENT > 512:
```
[the remark this answers is not in the document]
Comment (JeremieGauthier) [reply src/domain/models/bertsum_transformer.py:28]: J'approuve! Je vais faire le changement aussi.

File: src/domain/ucb.py (hunk 2 above)
Comment (JeremieGauthier) [reply src/domain/ucb.py:130]: I am happy you like it ;)

File: src/domain/models/bertsum_transformer.py (hunk 1 above)
Comment (JeremieGauthier) [reply src/domain/models/bertsum_transformer.py:52]: Not sure I'll use multi-GPU in CC anw.

File: src/domain/models/encoder.py [hunk 6]
```diff
@@ -80,7 +82,11 @@ def forward(self, top_vecs, mask):
...
 
         x = self.layer_norm(x)
-        sent_scores = self.sigmoid(self.wo(x))
-        sent_scores = sent_scores.squeeze(-1) * mask.float()
+        x = self.w0(x)
+        x = F.relu(x)
+        x = self.w1(x)
+
+        # sent_scores = self.sigmoid(x)
+        sent_scores = x.squeeze(-1) * mask.float()
```
[the remark this answers is not in the document]
Comment (JeremieGauthier) [reply src/domain/models/encoder.py:90]: Nous en avons déjà discuter sur le chat. Je garde ton commentaire en mémoire. Si nous avons des problèmes de performance, mais il me semble que j'avais des problèmes avec le vanishing

[the remark this answers is not in the document]
Comment (JeremieGauthier) [reply src/domain/models/encoder.py:90]: Je vais essayer à nouveau demain. Je vais faire les modifs au besoin

File: src/domain/models/bertsum_transformer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TortillasAlfred) [reply src/domain/models/bertsum_transformer.py:35]: Oui, mon commentaire est plutôt que ce serait plus clean et conforme d'utiliser D_MODEL_BERT que 768 directement.

[the remark this answers is not in the document]
Comment (JeremieGauthier) [reply src/domain/models/bertsum_transformer.py:35]: Je peux le changer.

File: src/domain/dataset_bert.py [hunk 7]
```diff
@@ -123,7 +135,11 @@ def tokenized_dataset(self, dataset):
         Returns:
             dict: dataset once tokenized
         """
-        tokenizer = BertTokenizerFast.from_pretrained("bert-base-uncased")
+        if self.config.bert_cache:  # Used if there's no internet connection
+            bert_cache_dir = join(getcwd(), "bert_cache/tokenizer_save_pretrained")
```
[the remark this answers is not in the document]
Comment (JeremieGauthier) [reply src/domain/dataset_bert.py:139]: Oui, tu as bien raison! Je viens de faire la correction. Donc dans la config, nous aurons string type pour notre cache path au lieu d'un bool type

File: src/domain/models/bertsum_transformer.py [hunk 8]
```diff
@@ -10,30 +10,32 @@
...
         self.device = device
-        if config.bert_cache:
-            bert_cache_dir = join(getcwd(), "bert_cache/bertmodel_save_pretrained")
-            self.bert = BertModel.from_pretrained(bert_cache_dir, local_files_only=True).to(self.device)
+        if not config.bert_cache:
+            self.bert = BertModel.from_pretrained("bert-base-uncased")
         else:
-            self.bert = BertModel.from_pretrained("bert-base-uncased").to(self.device)
+            self.bert = BertModel.from_pretrained(config.bert_cache, local_files_only=True)
+        self.bert.to(self.device)
```
Comment (TortillasAlfred) [inline src/domain/models/bertsum_transformer.py:28]:
```suggestion
        self.bert = self.bert.to(self.device)
```
Le .to() est pas inplace, il faut storer le modèle retourner par to()

File: src/domain/models/encoder.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (TortillasAlfred) [reply src/domain/models/encoder.py:89]: Confirmé: pas besoin du sigmoid dans la forward pass ici comme le BCELossWithLogits utilisée dans l'entraînement incorpore une sigmoide.

File: src/domain/dataset_bert.py [hunk 9]
```diff
@@ -112,7 +118,8 @@ def get_values(self, set_, part_, dataset, tokenizer):
         Returns:
             list: return a list of tokenized documents
         """
-        return [self.encode_document(document, tokenizer) for document in dataset[set_][part_]]
+        return Parallel(n_jobs=4)(delayed(encode_document)(document, tokenizer) for document in dataset[set_][part_])
```
Comment (TortillasAlfred) [inline src/domain/dataset_bert.py:121]:
```suggestion
        return Parallel(n_jobs=-1)(delayed(encode_document)(document, tokenizer) for document in dataset[set_][part_])
```

Comment (TortillasAlfred) [reply src/domain/dataset_bert.py]: What happens when encode_document encounters a None document (as in the None check it contains) ? 

Wouldn't this method then return a list possibly containing None ? i.e. [encoded_doc1, ..., None, ...] ? Doesn't this mess up the rest of the training process ? 

My personal suggestion:

`return Parallel(n_jobs=-1)(delayed(encode_document)(document, tokenizer) for document in dataset[set_][part_] if document)`

The `if document` filters out empty documents before parallelization to go around this possible bug.
